### ISSUE TITLE
Segmentation fault at ctranspose! after lufact! on OS X 10.10.1

### ISSUE THREAD

**Issue by rleegates** [JuliaLang/LinearAlgebra.jl#153]
Dear Julia developers,

I encountered a segmentation fault on a fresh install of OS X 10.10.1 and Julia 0.3.4 (problem also in 0.3.3) after a call to lufact! and \. The minimal example to reproducing this error is:

```
julia> A = speye(3)
3x3 sparse matrix with 3 Float64 entries:
    [1, 1]  =  1.0
    [2, 2]  =  1.0
    [3, 3]  =  1.0

julia> b = rand(3)
3-element Array{Float64,1}:
 0.932498
 0.340124
 0.613643

julia> lufact!(A)
UMFPACK LU Factorization of a 3-by-3 sparse matrix
Ptr{Void} @0x00007fd665312ba0


julia> A\b

signal (11): Segmentation fault: 11
ctranspose! at sparse/csparse.jl:190
ctranspose at sparse/csparse.jl:212
ishermitian at sparse/sparsematrix.jl:1779
CholmodSparse at linalg/cholmod.jl:458
factorize at linalg/cholmod.jl:1073
\ at linalg/generic.jl:233
jl_apply at /Users/vagrant/buildbot/slave/package_osx10_9/build/src/gf.c:1418
jl_apply at /Users/vagrant/buildbot/slave/package_osx10_9/build/src/interpreter.c:59
eval at /Users/vagrant/buildbot/slave/package_osx10_9/build/src/interpreter.c:207
jl_toplevel_eval_flex at /Users/vagrant/buildbot/slave/package_osx10_9/build/src/toplevel.c:496
jl_f_top_eval at /Users/vagrant/buildbot/slave/package_osx10_9/build/src/builtins.c:399
eval_user_input at REPL.jl:53
jlcall_eval_user_input_20042 at  (unknown line)
jl_apply at /Users/vagrant/buildbot/slave/package_osx10_9/build/src/./julia.h:981
anonymous at task.jl:95
jl_apply at /Users/vagrant/buildbot/slave/package_osx10_9/build/src/task.c:423
julia_trampoline at /Users/vagrant/buildbot/slave/package_osx10_9/build/src/init.c:1035
Segmentation fault: 11
```

My Julia versioninfo() gives:

```
Julia Version 0.3.4
Commit 3392026* (2014-12-26 10:42 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin13.4.0)
  CPU: Intel(R) Xeon(R) CPU           W3520  @ 2.67GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Nehalem)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

As I do not have a debug build, and the error is easy to reproduce, I hope this information will suffice.

Best regards,

Robert

**Comment by andreasnoack** [comment]
@rleegates The call `lufact!(A)` destroys `A` such that it is no longer a valid `SparseMatrixCSC` value. What you want to use for the solve is the output from `lufact!(A)`, i.e. 

```
Af = lufact!(A)
Af\b
```

That being said, it is quite unfortunate with a segfault. I'm not sure what's the best way of handling this. The reason for the segfault is that `lufact!` changes the sparse matrix to zero based indexing which is used in UMFPACK. A solution could be to restore valid `colptr` and `rowval` vectors at the end of `lufact!`. Another would be to check that at least `colptr[1]=rowval[1]=1` in all methods that use `@inbounds`.

Another thing, is it really the best way to check for Hermitianity as we do in

https://github.com/JuliaLang/julia/blob/master/base/sparse/sparsematrix.jl#L1908

It appears to me as an expensive computation.

cc: @ViralBShah, @tkelman

**Comment by tkelman** [comment]
I think last time `lufact!` for sparse matrices came up, I said that in-place factorizations for sparse matrices does not make sense, at least not with the exact same API as we use for dense. We have some of the symbolic/numeric functions exposed, but the `lufact!` style "in-place" functions use them somewhat awkwardly, and end up decrementing the indices to be zero-based as you said.

> Another thing, is it really the best way to check for Hermitianity as we do in
> 
> https://github.com/JuliaLang/julia/blob/master/base/sparse/sparsematrix.jl#L1908
> 
> It appears to me as an expensive computation.

No, you could definitely go through in one pass with code that looks a little like transpose but terminates early when it hits the first nonzero element in the upper triangle (assuming CSC here) that didn't have a matching nonzero in the lower triangle. No need to compute the entire transpose and subtraction.

**Comment by rleegates** [comment]
Andreas, thanks for the hint. This used to work on OS X 10.9, though. I believe I would have noticed if it gave the wrong results. What is the point in overwriting A then? 

On the Hermitian checking, why is it done at all? Usually, I know if my matrix is symmetric (I'm just working with reals) ahead of time. If it were symmetric (the real use-case isn't), I'd be using Cholesky instead of LU.

**Comment by rleegates** [comment]
On another note, perhaps this should be another issue: why is Julia using UMFPACK? Besides the restrictive GPL-licensing, it appears as though the MUMPS (public-domain) is the superior piece of the software. A comparison of MUMPS, PARDISO and UMFPACK can be found here: http://mumps.enseeiht.fr/doc/ud_2013/slides_rosseel.pdf

**Comment by andreasnoack** [comment]
The point is that you can use some of the memory allocated in `A` for the object returned form `lufact!`. However, as @tkelman it is a bit awkward right now and I think we'd be better off with not having the `!` versions for sparse matrices. The savings in allocation are not worth the confusion and risk of invalidating the input matrix.

I'd be surprised if this worked on 10.9. I cannot see how that should make a difference. Couldn't it be that you used `lufact!` in the way I did in the example instead?

The reason the matrix is checked for Hermitianity is that only the output from `lufact!` is a factorization type. The input remains a `SparseMatrixCSC` (although invalidated). Therefore when calling `A\b` after `lufact!(A)` you are asking for solving a general sparse linear system of equations and the matrix is therefore checked for symmetry to decide which factorization is the best to use.

**Comment by tkelman** [comment]
@rleegates there has been some work in interfacing to Mumps and various other sparse solvers, in several packages (https://github.com/JuliaSparse/MUMPSseq.jl and a few others, contributions to improve the packages are always welcome). There are a lot of options for sparse matrix codes, I've used almost all of them, and each has its own set of advantages and disadvantages - I wouldn't say any of them are absolutely uniformly superior in all cases. The reason SuiteSparse is included by default is because the wrappers to it were contributed early on, and it was incorporated into the build system. Eventually the idea would be to make a more flexible system that could use different backend solvers somewhat interchangeably, but it's going to take quite a bit of work to get there.

**Comment by rleegates** [comment]
@andreasnoack, okay, that makes more sense now :)

@tkelman, sounds like a good idea. I'm currently using https://github.com/dpo/MUMPS.jl

**Comment by ViralBShah** [comment]
@dmbates and I have also talked about writing a multifrontal sparse solver in Julia itself. There is no reason why we cannot make mumps the default solver, but it would need some benchmarking and thorough work on our part.

Because suitesparse solvers are used in Matlab, they tend to be rather well tested.

**Comment by ViralBShah** [comment]
We should also not have lufact! for all the reasons discussed.

**Comment by rleegates** [comment]
@ViralBShah How would a multi-frontal solver fare in a GC world? On benchmarking, I have attached a link to a comparison of the solvers UMFPACK, PARDISO, and MUMPS above. I doubt that MUMPS can be considered to be not as well tested if companies such as MSC use them as their default solver. Here's a provocative one: does Julia want to be MATLAB? :)

Just to give a real-world, motivating example: I am solving a large contact problem (variational inequality) of approx. 780k DOFs (reverse Cuthill-McKee ordering). UMFPACK uses >50GB of memory for this problem, lacking an out-of-core option. I have a fairly capable machine, however my 32 GB of memory are quickly exhausted here (even with memory compression, process exits due to insufficient memory), so an OOC option is vital. Right now, such an option is out-of-reach without resorting to external packages.

**Comment by tkelman** [comment]
@rleegates I have my own benchmark data that shows MUMPS performing quite poorly when compared to HSL and WSMP solvers on nonlinear optimization problems, it all depends on your problem size, structure, and application area. Mumps has a really awful build system (SuiteSparse isn't a whole lot better here, but it is better), it's in Fortran and we'd rather get rid of our Fortran dependencies than add any more, and it can't be used in parallel without MPI. By all means try to make the Mumps packages better (packages are not going to be promoted to base functionality until they work well across all platforms and are thoroughly tested - none of the Mumps packages are even close here), but in my experience Mumps isn't worth the trouble. Out-of-core solvers are an interesting topic and vital for some types of work, but I don't see it as something that needs to be provided by the standard library any time soon. We'll probably move SuiteSparse out into its own package at some point, but probably only when there are sufficiently general frameworks in place for sparse linear algebra to make it possible for packages to supply different backends for the same high-level syntax.

**Comment by rleegates** [comment]
@tkelman I'm always open to new ideas, and I haven't done any extensive benchmarking myself (just noticed 50% speed-up over UMFPACK), so I cannot comment in a qualified manner on the topic of efficiency for all use-cases. Lacking a numerical mathematics background, my knowledge as to the inner workings of these pieces of software is very limited, so it'd probably be better for everyone if I'd refrain from contributing to these packages. I was just trying to convey an end-user's opinion :)

**Comment by dpo** [comment]
SuiteSparse is a great option for a general-purpose factorization I think, and it will also do sparse Cholesky and sparse QR. The latter is particularly useful for least-squares problems. But when it comes to symmetric indefinite problems, you'll want to use a symmetric indefinite factorization, which SuiteSparse doesn't offer (with the exception of very specific cases where it's possible to show that a Cholesky-like factorization exists and is stable and CholMod will do a good job). Now in the category of open-source symmetric indefinite factorization, I only know of MUMPS. Yes, the HSL solvers sometimes perform better but they have restrictive licensing. WSMP is commercial-ish, as is PARDISO. Now that Homebrew supplies bottles for MUMPS, and thanks to MPI.jl, it's become a viable option for end users in my opinion.

**Comment by tkelman** [comment]
>  the HSL solvers sometimes perform better

Almost always, in my experience. But again, problem dependent, so perhaps your experience differs.

> Now in the category of open-source symmetric indefinite factorization, I only know of MUMPS.

Elemental has some relatively new sparse factorization functionality that should be usable and hopefully competitive, though I admittedly haven't benchmarked it yet and it shares some difficulties in common with Mumps of relying solely on MPI for parallelization.

> it's become a viable option for end users in my opinion

For users on OSX, perhaps. Packages should really attempt to follow best practices and support non-OSX platforms. Also licensing of a Julia package as GPL when it's wrapping more permissive code is problematic for much of the community.

**Comment by rleegates** [comment]
I can definitely second Tony's opinion on the GPL.

**Comment by dpo** [comment]
It's true that that's inconsistent. I think I mechanically put the GPL license in there without checking. MUMPS isn't using much of a license. What would be the official open-source license closest to the MUMPS terms? MIT? Unlicense?

Re supporting other platforms: I haven't touched a Windows machine in years and know nothing about compilation there. Pull requests welcome. If Homebrew is able to install MUMPS on Windows and Linux, and Julia holds up to its portability promises, I don't see why having `MUMPS.jl` running on those platforms should be a problem.

**Comment by rleegates** [comment]
MUMPS is in the public domain. No license would basically imply a copyright.

http://mumps.enseeiht.fr/index.php?page=credits

This version of MUMPS is provided to you free of charge. It is public domain, based on public domain software developed during the Esprit IV European project PARASOL (1996-1999). Since this first public domain version in 1999, research and developments have been supported by the following institutions: CERFACS, CNRS, ENS Lyon, INPT(ENSEEIHT)-IRIT, INRIA, and University of Bordeaux. Part of MUMPS related work has been funded by the following contracts: contracts with Samtech S.A. (2005-2006, 2008-2010); contract with CNES-CERFACS (2005), ANR project SOLSTICE ANR-06-CIS6-010 (2007-2009) labelled by A.E.S.E (Aerospace Vallée) competitivity pole; France-Berkeley projects (1999-2000 and 2008-2009); France-Israel technological cooperation programs on Multicomputing: Multicore, Cluster and Grid (2009-2010); EDF grant for the PhD thesis of C. Weisbecker (2010-2012); contract with TOTAL-Hutchinson (2013), contract with ESI group (2013).

**Comment by dpo** [comment]
And Julia itself is MIT, so MIT might make the most sense.

**Comment by andreasnoack** [comment]
Fix by JuliaLang/LinearAlgebra.jl#153.
